driver/opc/writer.h

```diff
@@ -0,0 +1,204 @@
+//                                    writer config                              //
+///////////////////////////////////////////////////////////////////////////////////
+struct WriterConfig {
+    /// @brief the device representing the OPC UA server to read from.
+    std::string device;
+    /// @brief the list of channels to read from the server.
+    std::vector<WriterChannelConfig> channels;
+    /// @brief frequency state of a controlled channel is published
+    synnax::Rate state_rate = synnax::Rate(1); // default to 1 Hz
+    /// @brief index key for all state channels in this task
+    synnax::ChannelKey state_index_key;
+
+    WriterConfig() = default;
+
+    explicit WriterConfig(config::Parser &parser);
+
+    [[nodiscard]] std::vector<ChannelKey> cmd_keys() const {
+        std::vector<ChannelKey> keys;
+        for (const auto &channel: channels) keys.push_back(channel.cmd_channel);
+        return keys;
+    }
+}; // struct WriterConfig
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                    OPC Sink                                   //
+///////////////////////////////////////////////////////////////////////////////////
+/// @brief an OPC writer with embedded OPC UA client that receives data from synnax
+/// in frames, and writes them to the appropriate nodes on the connected OPC UA Server.
+class WriterSink final : public pipeline::Sink {
+public:
+    // Synnax Resources
+    std::shared_ptr<task::Context> ctx;
+    synnax::Task task;
+    std::map<ChannelKey, WriterChannelConfig> cmd_channel_map; // TODO: Change to cmd channel map
+    task::State curr_state;
+
+    // OPC UA Resources
+    WriterConfig cfg;
+    std::shared_ptr<UA_Client> ua_client;
+    UA_WriteRequest req; // defined in types_generated.h
+    opc::DeviceProperties device_props;
+
+    // keep alive resources (thread and mutex)
+    std::thread keep_alive_thread;
+    std::mutex client_mutex;
+    breaker::Breaker breaker;
+    ///@brief the rate at which sink will ping the OPC UA server to maintain the connection
+    synnax::Rate ping_rate = synnax::Rate(0.1); // default to every 10s
+
+    WriterSink(
+        WriterConfig cfg,
+        const std::shared_ptr<UA_Client> &ua_client,
+        const std::shared_ptr<task::Context> &ctx,
+        synnax::Task task,
+        opc::DeviceProperties device_props
+    );
+
+    ~WriterSink() override {
+        this->breaker.stop();
+        this->keep_alive_thread.join();
+    }
+
+    freighter::Error write(synnax::Frame frame) override;
+
+    void maintain_connection();
+
+private:
+    void initialize_write_request(const synnax::Frame &frame);
+
+    void initialize_write_value(
+        const synnax::Frame &frame,
+        uint32_t &index,
+        WriterChannelConfig &ch,
+        UA_WriteValue *write_value
+    );
+
+    void stoppedWithErr(const freighter::Error &err) override;
```

Comment:
We should just go ahead and fix the naming convention on this function now. `stopped_with_err`

Reply:
done